### PR TITLE
Add memory-efficient fused linear cross-entropy loss (solves issue #35906)

### PR DESCRIPTION
**Pull request by Coffeempty**
Closes #35906 

## What
 
Adds `jax.experimental.fused_linear_cross_entropy_loss` — computes `cross_entropy(x @ w.T, labels)` without materializing the full `[N, V]` logits tensor.
 
## Why
 
For large vocabularies (128K+ tokens in Llama-3, 202K in Llama-4), the intermediate logits tensor dominates training memory. At typical LLM training scale:
 
| Config | Full logits | Chunked (cs=4096) | Reduction |
|---|---|---|---|
| Llama-3 (V=128K, N=8192, bf16) | 4.0 GB | 128 MB | 97% |
| Llama-4 (V=202K, N=8192, bf16) | 6.3 GB | 128 MB | 98% |
 
This is a solved problem in PyTorch (Apple CCE, Liger-Kernel) but JAX has no equivalent.
 
## How
 
Based on "Cut Your Losses in Large-Vocabulary Language Models" ([Wijmans et al., 2024](https://arxiv.org/abs/2411.09009)):
 
- **Forward:** Computes target logit directly in O(N·D). Log-sum-exp denominator computed via `lax.scan` over vocabulary chunks with online max + sum(exp) accumulators.
- **Backward:** `custom_vjp` recomputes logits per-chunk (trading compute for memory). Accumulates `dx` and `dw` without ever storing `[N, V]` tensors. Only residuals saved: `x`, `w`, and per-position `lse` (shape `[N]`).
- Pure JAX — no Pallas or Triton dependency. Works on CPU, GPU, and TPU.
 
## Tests
 
33 tests covering:
 
- Forward accuracy vs naive full-matmul implementation (float32 + bfloat16, multiple shapes)
- Gradient accuracy for `dx` and `dw`
- `jit` compilation (forward + backward)
- Edge cases: V=1, V < chunk\_size, chunk\_size=1, N=1, large vocab, non-divisible sizes
- Invalid input validation
 
## Disclosure
 
This implementation was developed with AI assistance (Claude). I have reviewed all code, verified correctness across all test cases, and understand the algorithm and design decisions. Happy to discuss any part of the implementation in review.

cc @pillow37

### CODE REVIEW

**Comment by pillow37** [comment]
Hi @Coffeempty
thank you so much for implementing this and so quickly.

I would have a few improvement suggestions before the PR goes through:

1) In the [CCE implementation](https://github.com/apple/ml-cross-entropy?tab=readme-ov-file#getting-started) they provide the "shift_embeddings" kwarg which avoids materialization of the following. Would it be easily possible to implement this?
```
shift_embeddings = embeddings[..., :-1, :].flatten(0, -2)
shift_labels = labels[..., 1:]
```

2) Is it possible to donate the buffer from the pre-logits for their gradients? In the paper they mention that they do this. But I'm not sure if this is possible in a jax implementation where fwd and bwd are separated.

3) There are several possible improvement. The paper impressively shows the effect of vocabulary sorting and gradient filtering. Any ideas how to implement those?
image

**Comment by Coffeempty** [comment]
heyy @pillow37 

Thanks for the detailed feedback! All great suggestions.

1. shift kwarg : will add this, straightforward.
2. Buffer donation : will look into what's possible within JAX's functional model. Since we never materialize full logits, the savings may already be implicit, but I'll see if donate_argnums can help.
3. Vocab sorting + gradient filtering : will take a shot at both.

**Comment by Coffeempty** [comment]
hey @pillow37 

Pushed updates for all three points, 51 tests passing now:
1. shift parameter - Added a shift=0 kwarg. Works with both flat [T, D] and batched [B, T, D] inputs. Setting shift=1 covers the standard causal LM case without needing to materialize the shifted tensors yourself.
2. Buffer donation - Looked into this. JAX's XLA backend already reuses buffer memory automatically through liveness analysis, so since our forward residuals get consumed during backward, XLA can reclaim that memory on its own. Users can also pass donate_argnums on their outer jax.jit call for more explicit control. There's no direct equivalent to the in-place overwrite trick from the Triton kernel, but we never materialize full logits anyway so the memory pressure is already pretty minimal. Added a comment in the code explaining this.
3. Vocab sorting + gradient filtering - Added both as optional params. vocab_sort_indices reorders weight rows before chunking, and gradients get unscattered back to the original order automatically through JAX's lax.gather VJP. filter_eps zeros out chunks in the backward pass when their max logit is far below the global max (uses masking instead of branching to stay compatible with lax.scan). Forward pass stays exact regardless.

Let me know if this lines up with what you were thinking or if you'd want anything changed.

**Comment by jakevdp** [comment]
Hi - thanks for this! At first glance the code looks fine: that said, I'm not sure the JAX team can commit to maintaining this. Maybe you could create a new package with this tool in it?

**Comment by Coffeempty** [comment]
Hi @jakevdp
Makes total sense, thanks for the review. Published it as a standalone package: https://pypi.org/project/jax-cce/
Repo: https://github.com/Coffeempty/jax-cce
